___

<p style="text-align: center;"><img src="https://docs.google.com/uc?id=1lY0Uj5R04yMY3-ZppPWxqCr5pvBLYPnV" class="img-fluid" 
alt="CLRSWY"></p>

## <p style="background-color:#FDFEFE; font-family:newtimeroman; color:#9d4f8c; font-size:100%; text-align:center; border-radius:10px 10px;">WAY TO REINVENT YOURSELF</p>

## <p style="background-color:#FDFEFE; font-family:newtimeroman; color:#060108; font-size:200%; text-align:center; border-radius:10px 10px;">DA & DVwPY</p>

## <p style="background-color:#FDFEFE; font-family:newtimeroman; color:#060108; font-size:200%; text-align:center; border-radius:10px 10px;">The Exploratory Data Analysis (EDA) Project</p>

<img src=https://i.ibb.co/wJW61Y2/Used-cars.jpg width="700" height="200">

## <p style="background-color:#FDFEFE; font-family:newtimeroman; color:#060108; font-size:200%; text-align:center; border-radius:10px 10px;">AutoScout Car Price Prediction EDA</p>



## Introduction
Welcome to "***AutoScout Exploratory Data Analysis (EDA) Project***". This is the project of completing ***Data Analysis & Data Visualization*** Courses and a successful hand-over is mandatory for getting certification for both courses. **Auto Scout** data used in this project scraped from the Website of an online car trading company in 2022, and contains many features of 13 different car makes including 594 models. In this project, you will have the opportunity to apply many commonly used algorithms for Data Cleaning and Exploratory Data Analysis by using a variety of Python libraries, such as Numpy, Pandas, Matplotlib, Seaborn, Scipy, and then you will get a clean dataset for your analysis and pretictive modelling in Machine Learning Path. So you will have the chance to use all the skills you have already learned in the Data Analysis and Visualization courses.

**``In this context, the project consists of 3 parts in general:``**
* **The first part** is related to ``'Data Cleaning'``. It deals with Incorrect Headers, Incorrect Format, Anomalies, and Dropping useless columns.
* **The second part** is related to ``'Filling Data'``, in other words 'Imputation'. It deals with Missing Values. Categorical to numeric transformation, Encoding, is done as well.
* **The third part** is related to ``'Handling Outliers of Data'`` via Visualization libraries. So, some insights will be extracted.

**``SPECIAL NOTE:``**  However, you are ``free to create your own style``. You do ``NOT`` have to stick to the steps above. Nevertheless, we, the DA & DV instructors, recommend you study each part separately to create a source notebook for your further studies. 

In order to build your Portfolio in terms of the GitHub account, you need to push your solution file up to your own repository.

Please after solving the task, commit your notebook to GitHub and submit its link to LMS page where the project is settled down.

**BE NOTED:** Please ``DO NOT FORGET`` to click the submit button.

### Some Reminders on Exploratory data analysis (EDA)

Exploratory data analysis (EDA) is an especially important activity in the routine of a data analyst or scientist. It enables an in depth understanding of the dataset, define or discard hypotheses and create predictive models on a solid basis. It uses data manipulation techniques and several statistical tools to describe and understand the relationship between variables and how these can impact business. By means of EDA, we can obtain meaningful insights that can impact analysis under the following questions (If a checklist is good enough for pilots to use every flight, it’s good enough for data scientists to use with every dataset).
1. What question are you trying to solve (or prove wrong)?
2. What kind of data do you have?
3. What’s missing from the data?
4. Where are the outliers?
5. How can you add, change or remove features to get more out of your data?

**``Exploratory data analysis (EDA)``** is often an **iterative brainstorming process** where you pose a question, review the data, and develop further questions to investigate before beginning model development work. The image below shows how the brainstorming phase is connected with that of understanding the variables and how this in turn is connected again with the brainstorming phase.<br>

<img src=https://i.ibb.co/k0MC950/EDA-Process.png width="300" height="100">

[Image Credit: Andrew D.](https://towardsdatascience.com/exploratory-data-analysis-in-python-a-step-by-step-process-d0dfa6bf94ee)



In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import re

In [2]:
df = pd.read_json('as24_cars.json')
df

make_model  \
0         Mercedes-Benz A 160   
1       Mercedes-Benz EQE 350   
2      Mercedes-Benz A 45 AMG   
3      Mercedes-Benz A 35 AMG   
4      Mercedes-Benz A 45 AMG   
...                       ...   
29475               Volvo S40   
29476               Volvo C30   
29477               Volvo S90   
29478               Volvo S80   
29479                     NaN   

                                       short_description               make  \
0                                                    CDi  \nMercedes-Benz\n   
1                                                   350+  \nMercedes-Benz\n   
2                                       S 4Matic+ 8G-DCT  \nMercedes-Benz\n   
3                                         4Matic+ 7G-DCT  \nMercedes-Benz\n   
4                               200CDI BE Line 4M 7G-DCT  \nMercedes-Benz\n   
...                                                  ...                ...   
29475        2.4i Momentum/ Airco/ NAP/ APK/ Elek. Ramen          \nVolvo\n   
29476                    1.6 D DRIVe Start/Stop R-design          \nVolvo\n   
29477  T5 AUT(8) MOMENTUM BUSINESS SCHUIFDAK 360GR CA...          \nVolvo\n   
29478  2.9 Executive G. NETTE AUTO! LEER! NAVI! CRUIS...          \nVolvo\n   
29479                                                NaN                NaN   

                    model                                           location  \
0         [\n, A 160 ,\n]  P.I. EL PALMAR C/FORJA 6,  11500 PUERTO DE SAN...   
1       [\n, EQE 350 ,\n]       APARTADO DE CORREOS 1032,  26140 LOGROÑO, ES   
2      [\n, A 45 AMG ,\n]  PORT. TARRACO, MOLL DE LLEVANT, Nº 5, LOC. 6-8...   
3      [\n, A 35 AMG ,\n]  Carrer de Provença, 31 Local,  8029 BARCELONA, ES   
4      [\n, A 45 AMG ,\n]             CARRIL ARAGONES 4,  30007 CASILLAS, ES   
...                   ...                                                ...   
29475       [\n, S40 ,\n]                  Stobbeweg 6,  2461 EX TER AAR, NL   
29476       [\n, C30 ,\n]             Sur rendez-vous,  5060 Sambreville, BE   
29477       [\n, S90 ,\n]    Bergambachterstraat 5,  2871 JB SCHOONHOVEN, NL   
29478       [\n, S80 ,\n]                  Lage Ham 106,  5102 AE DONGEN, NL   
29479                 NaN                                                NaN   

            price          Body type                      Type        Doors  \
0      € 16,950.-  [\n, Compact, \n]            [\n, Used, \n]  [\n, 5, \n]   
1      € 80,900.-  [\n, Compact, \n]  [\n, Pre-registered, \n]  [\n, 4, \n]   
2      € 69,900.-  [\n, Compact, \n]            [\n, Used, \n]  [\n, 5, \n]   
3      € 46,990.-  [\n, Compact, \n]            [\n, Used, \n]  [\n, 5, \n]   
4      € 16,800.-  [\n, Compact, \n]            [\n, Used, \n]  [\n, 5, \n]   
...           ...                ...                       ...          ...   
29475   € 5,499.-    [\n, Sedan, \n]            [\n, Used, \n]  [\n, 4, \n]   
29476   € 7,300.-    [\n, Sedan, \n]            [\n, Used, \n]  [\n, 3, \n]   
29477  € 29,900.-    [\n, Sedan, \n]            [\n, Used, \n]  [\n, 4, \n]   
29478   € 4,440.-    [\n, Sedan, \n]            [\n, Used, \n]  [\n, 4, \n]   
29479         NaN                NaN                       NaN          NaN   

         Country version  ... CO₂-efficiency Fuel consumption (WLTP)  \
0        [\n, Spain, \n]  ...            NaN                     NaN   
1        [\n, Spain, \n]  ...            NaN                     NaN   
2        [\n, Spain, \n]  ...            NaN                     NaN   
3        [\n, Spain, \n]  ...            NaN                     NaN   
4        [\n, Spain, \n]  ...            NaN                     NaN   
...                  ...  ...            ...                     ...   
29475                NaN  ...            NaN                     NaN   
29476  [\n, Belgium, \n]  ...            NaN                     NaN   
29477                NaN  ...            NaN                     NaN   
29478                NaN  ...            Na

# PART- 1 `( Data Cleaning )`

## <p style="background-color:yellow; font-family:newtimeroman; color:blue; font-size:150%; text-align:center; border-radius:10px 10px;">Renaming of columns</p> 

In [3]:
df.columns

Index(['make_model', 'short_description', 'make', 'model', 'location', 'price',
       'Body type', 'Type', 'Doors', 'Country version', 'Offer number',
       'Warranty', 'Mileage', 'First registration', 'Gearbox', 'Fuel type',
       'Colour', 'Paint', 'desc', 'seller', 'Seats', 'Power', 'Engine size',
       'Gears', 'CO₂-emissions', 'Manufacturer colour', 'Drivetrain',
       'Cylinders', 'Fuel consumption', '\nComfort & Convenience\n',
       '\nEntertainment & Media\n', '\nSafety & Security\n', '\nExtras\n',
       'Empty weight', 'Model code', 'General inspection', 'Last service',
       'Full service history', 'Non-smoker vehicle', 'Emission class',
       'Emissions sticker', 'Upholstery colour', 'Upholstery',
       'Production date', 'Previous owner', 'Other fuel types',
       'Power consumption', 'Energy efficiency class', 'CO₂-efficiency',
       'Fuel consumption (WLTP)', 'CO₂-emissions (WLTP)', 'Available from',
       'Taxi or rental car', 'Availability', 'Last timing b

In [4]:
df.columns = df.columns.str.strip()  

def modify_column_name(column_name):
    modified_name = re.sub(r'\s+', '_', column_name.strip())  
    return modified_name.title()  

df.columns = df.columns.map(modify_column_name) 
df.columns

Index(['Make_Model', 'Short_Description', 'Make', 'Model', 'Location', 'Price',
       'Body_Type', 'Type', 'Doors', 'Country_Version', 'Offer_Number',
       'Warranty', 'Mileage', 'First_Registration', 'Gearbox', 'Fuel_Type',
       'Colour', 'Paint', 'Desc', 'Seller', 'Seats', 'Power', 'Engine_Size',
       'Gears', 'Co₂-Emissions', 'Manufacturer_Colour', 'Drivetrain',
       'Cylinders', 'Fuel_Consumption', 'Comfort_&_Convenience',
       'Entertainment_&_Media', 'Safety_&_Security', 'Extras', 'Empty_Weight',
       'Model_Code', 'General_Inspection', 'Last_Service',
       'Full_Service_History', 'Non-Smoker_Vehicle', 'Emission_Class',
       'Emissions_Sticker', 'Upholstery_Colour', 'Upholstery',
       'Production_Date', 'Previous_Owner', 'Other_Fuel_Types',
       'Power_Consumption', 'Energy_Efficiency_Class', 'Co₂-Efficiency',
       'Fuel_Consumption_(Wltp)', 'Co₂-Emissions_(Wltp)', 'Available_From',
       'Taxi_Or_Rental_Car', 'Availability', 'Last_Timing_Belt_Change',
   

In [5]:
df.head(1)

Make_Model Short_Description               Make            Model  \
0  Mercedes-Benz A 160               CDi  \nMercedes-Benz\n  [\n, A 160 ,\n]   

                                            Location       Price  \
0  P.I. EL PALMAR C/FORJA 6,  11500 PUERTO DE SAN...  € 16,950.-   

           Body_Type            Type        Doors  Country_Version  ...  \
0  [\n, Compact, \n]  [\n, Used, \n]  [\n, 5, \n]  [\n, Spain, \n]  ...   

  Co₂-Efficiency Fuel_Consumption_(Wltp) Co₂-Emissions_(Wltp) Available_From  \
0            NaN                     NaN                  NaN            NaN   

  Taxi_Or_Rental_Car Availability Last_Timing_Belt_Change  \
0                NaN          NaN                     NaN   

  Electric_Range_(Wltp) Power_Consumption_(Wltp) Battery_Ownership  
0                   NaN                      NaN               NaN  

[1 rows x 58 columns]

*I created 2 copy of my df to avoid confusion

In [6]:
df1=df.copy()
df0=df.copy()

## <p style="background-color:yellow; font-family:newtimeroman; color:blue; font-size:150%; text-align:center; border-radius:10px 10px;">Reorganising dirty columns</p> 

In [7]:
df.iloc[1].value_counts

<bound method IndexOpsMixin.value_counts of Make_Model                                              Mercedes-Benz EQE 350
Short_Description                                                        350+
Make                                                        \nMercedes-Benz\n
Model                                                       [\n, EQE 350 ,\n]
Location                         APARTADO DE CORREOS 1032,  26140 LOGROÑO, ES
Price                                                              € 80,900.-
Body_Type                                                   [\n, Compact, \n]
Type                                                 [\n, Pre-registered, \n]
Doors                                                             [\n, 4, \n]
Country_Version                                               [\n, Spain, \n]
Offer_Number                                                [\n, 7054863, \n]
Warranty                                                  [\n, 24 months, \n]
Mileage             

** <b>Column name of 'Make'<b>

In [8]:
df['Make'] = df['Make'].apply(lambda item: re.sub(r'[\n,]', '', item) if type(item) == str else item)
df['Make']

0        Mercedes-Benz
1        Mercedes-Benz
2        Mercedes-Benz
3        Mercedes-Benz
4        Mercedes-Benz
             ...      
29475            Volvo
29476            Volvo
29477            Volvo
29478            Volvo
29479              NaN
Name: Make, Length: 29480, dtype: object

** <b>Column name of 'Model'<b>

In [9]:
df['Model'] = df['Model'].apply(lambda item: ",".join(item) if type(item) == list else item)
df['Model']

0           \n, A 160 ,\n
1         \n, EQE 350 ,\n
2        \n, A 45 AMG ,\n
3        \n, A 35 AMG ,\n
4        \n, A 45 AMG ,\n
               ...       
29475         \n, S40 ,\n
29476         \n, C30 ,\n
29477         \n, S90 ,\n
29478         \n, S80 ,\n
29479                 NaN
Name: Model, Length: 29480, dtype: object

In [10]:
df['Model'] = df['Model'].apply(lambda item: re.sub(r'[\n, ]', '', item) if type(item) == str else item)
df['Model']

0          A160
1        EQE350
2        A45AMG
3        A35AMG
4        A45AMG
          ...  
29475       S40
29476       C30
29477       S90
29478       S80
29479       NaN
Name: Model, Length: 29480, dtype: object

** <b>Column name of 'Location'<b>
* I filled that column with just the country codes and null values

In [11]:
df['Location']

0        P.I. EL PALMAR C/FORJA 6,  11500 PUERTO DE SAN...
1             APARTADO DE CORREOS 1032,  26140 LOGROÑO, ES
2        PORT. TARRACO, MOLL DE LLEVANT, Nº 5, LOC. 6-8...
3        Carrer de Provença, 31 Local,  8029 BARCELONA, ES
4                   CARRIL ARAGONES 4,  30007 CASILLAS, ES
                               ...                        
29475                    Stobbeweg 6,  2461 EX TER AAR, NL
29476               Sur rendez-vous,  5060 Sambreville, BE
29477      Bergambachterstraat 5,  2871 JB SCHOONHOVEN, NL
29478                    Lage Ham 106,  5102 AE DONGEN, NL
29479                                                  NaN
Name: Location, Length: 29480, dtype: object

In [12]:
for column in df:
    df['Location'] = df['Location'].str.extract(r'(\b\w{2}$)').fillna(np.nan)
    
df['Location']

0         ES
1         ES
2         ES
3         ES
4         ES
        ... 
29475     NL
29476     BE
29477     NL
29478     NL
29479    NaN
Name: Location, Length: 29480, dtype: object

** <b>Column name of 'Price'<b>

* I dropped sign of '$ , . - and spaces' than changed the type to float

In [13]:
df['Price']

0        € 16,950.-
1        € 80,900.-
2        € 69,900.-
3        € 46,990.-
4        € 16,800.-
            ...    
29475     € 5,499.-
29476     € 7,300.-
29477    € 29,900.-
29478     € 4,440.-
29479           NaN
Name: Price, Length: 29480, dtype: object

In [14]:
df['Price'] = df['Price'].str.extract(r'(\d+\,?\d*)')
df['Price'] = df['Price'].str.replace(',',"").astype(float)
df['Price']

0        16950.0
1        80900.0
2        69900.0
3        46990.0
4        16800.0
          ...   
29475     5499.0
29476     7300.0
29477    29900.0
29478     4440.0
29479        NaN
Name: Price, Length: 29480, dtype: float64

** <b> Column name of 'Body_Type'<b>

In [15]:
df['Body_Type'] = df['Body_Type'].apply(lambda item: ",".join(item) if type(item) == list else item)
df['Body_Type'] 

0        \n, Compact, \n
1        \n, Compact, \n
2        \n, Compact, \n
3        \n, Compact, \n
4        \n, Compact, \n
              ...       
29475      \n, Sedan, \n
29476      \n, Sedan, \n
29477      \n, Sedan, \n
29478      \n, Sedan, \n
29479                NaN
Name: Body_Type, Length: 29480, dtype: object

In [16]:
df['Body_Type'] = df['Body_Type'].apply(lambda item: re.sub(r'[\n,]', '', item) if type(item) == str else item)
df['Body_Type']

0         Compact 
1         Compact 
2         Compact 
3         Compact 
4         Compact 
           ...    
29475       Sedan 
29476       Sedan 
29477       Sedan 
29478       Sedan 
29479          NaN
Name: Body_Type, Length: 29480, dtype: object

** <b> Column name of 'Type'<b>

In [17]:
df['Type'] 

0                  [\n, Used, \n]
1        [\n, Pre-registered, \n]
2                  [\n, Used, \n]
3                  [\n, Used, \n]
4                  [\n, Used, \n]
                   ...           
29475              [\n, Used, \n]
29476              [\n, Used, \n]
29477              [\n, Used, \n]
29478              [\n, Used, \n]
29479                         NaN
Name: Type, Length: 29480, dtype: object

In [18]:
df['Type'] = df['Type'].apply(lambda item: ",".join(item) if type(item) == list else item)
df['Type']

0                  \n, Used, \n
1        \n, Pre-registered, \n
2                  \n, Used, \n
3                  \n, Used, \n
4                  \n, Used, \n
                  ...          
29475              \n, Used, \n
29476              \n, Used, \n
29477              \n, Used, \n
29478              \n, Used, \n
29479                       NaN
Name: Type, Length: 29480, dtype: object

In [19]:
df['Type'] = df['Type'].apply(lambda item: re.sub(r'[\n,]', '', item) if type(item) == str else item)
df['Type']

0                   Used 
1         Pre-registered 
2                   Used 
3                   Used 
4                   Used 
               ...       
29475               Used 
29476               Used 
29477               Used 
29478               Used 
29479                 NaN
Name: Type, Length: 29480, dtype: object

** <b> Column name of 'Doors'<b>

In [20]:
df['Doors']

0        [\n, 5, \n]
1        [\n, 4, \n]
2        [\n, 5, \n]
3        [\n, 5, \n]
4        [\n, 5, \n]
            ...     
29475    [\n, 4, \n]
29476    [\n, 3, \n]
29477    [\n, 4, \n]
29478    [\n, 4, \n]
29479            NaN
Name: Doors, Length: 29480, dtype: object

In [21]:
df['Doors'] = df['Doors'].apply(lambda item: ",".join(item) if type(item) == list else item)
df['Doors']

0        \n, 5, \n
1        \n, 4, \n
2        \n, 5, \n
3        \n, 5, \n
4        \n, 5, \n
           ...    
29475    \n, 4, \n
29476    \n, 3, \n
29477    \n, 4, \n
29478    \n, 4, \n
29479          NaN
Name: Doors, Length: 29480, dtype: object

In [22]:
df['Doors'] = df['Doors'].apply(lambda item: re.sub(r'[\n,]', '', item) if type(item) == str else item)
df['Doors']

0         5 
1         4 
2         5 
3         5 
4         5 
        ... 
29475     4 
29476     3 
29477     4 
29478     4 
29479    NaN
Name: Doors, Length: 29480, dtype: object

** <b>Column name of 'Country_Version'<b>

In [23]:
df['Country_Version']

0          [\n, Spain, \n]
1          [\n, Spain, \n]
2          [\n, Spain, \n]
3          [\n, Spain, \n]
4          [\n, Spain, \n]
               ...        
29475                  NaN
29476    [\n, Belgium, \n]
29477                  NaN
29478                  NaN
29479                  NaN
Name: Country_Version, Length: 29480, dtype: object

In [24]:
df['Country_Version']=df['Country_Version'].apply(lambda item: ",".join(item) if type(item) == list else item)
df['Country_Version']

0          \n, Spain, \n
1          \n, Spain, \n
2          \n, Spain, \n
3          \n, Spain, \n
4          \n, Spain, \n
              ...       
29475                NaN
29476    \n, Belgium, \n
29477                NaN
29478                NaN
29479                NaN
Name: Country_Version, Length: 29480, dtype: object

In [25]:
df['Country_Version'] = df['Country_Version'].apply(lambda item: re.sub(r'[\n,]', '', item) if type(item) == str else item)
df['Country_Version']

0           Spain 
1           Spain 
2           Spain 
3           Spain 
4           Spain 
           ...    
29475          NaN
29476     Belgium 
29477          NaN
29478          NaN
29479          NaN
Name: Country_Version, Length: 29480, dtype: object

** <b>Column name of 'Offer_Number'<b>

In [26]:
df['Offer_Number']

0         [\n, 7468157, \n]
1         [\n, 7054863, \n]
2         [\n, 7410454, \n]
3         [\n, 7464503, \n]
4         [\n, 7127931, \n]
                ...        
29475    [\n, 09-PF-XS, \n]
29476                   NaN
29477        [\n, 2791, \n]
29478    [\n, 43-JR-LR, \n]
29479                   NaN
Name: Offer_Number, Length: 29480, dtype: object

In [27]:
df['Offer_Number']=df['Offer_Number'].apply(lambda item: ",".join(item) if type(item) == list else item)
df['Offer_Number']

0         \n, 7468157, \n
1         \n, 7054863, \n
2         \n, 7410454, \n
3         \n, 7464503, \n
4         \n, 7127931, \n
               ...       
29475    \n, 09-PF-XS, \n
29476                 NaN
29477        \n, 2791, \n
29478    \n, 43-JR-LR, \n
29479                 NaN
Name: Offer_Number, Length: 29480, dtype: object

In [28]:
df['Offer_Number'] = df['Offer_Number'].apply(lambda item: re.sub(r'[\n,]', '', item) if type(item) == str else item)
df['Offer_Number']

0          7468157 
1          7054863 
2          7410454 
3          7464503 
4          7127931 
            ...    
29475     09-PF-XS 
29476           NaN
29477         2791 
29478     43-JR-LR 
29479           NaN
Name: Offer_Number, Length: 29480, dtype: object

In [29]:
df['Offer_Number'].is_unique

False

In [30]:
df['Offer_Number'].value_counts()

 1                         28
 L-Vorlauf 2023            10
 20                         9
 30                         9
 RE82542                    8
                           ..
 ggp-EP-880-ZS              1
 7475319                    1
 abci-EZ-260-PS_130291      1
 7407611                    1
 43-JR-LR                   1
Name: Offer_Number, Length: 20945, dtype: int64

** <b>Column name of 'Warranty'<b>

In [31]:
df['Warranty']

0        [\n, 24 months, \n]
1        [\n, 24 months, \n]
2        [\n, 12 months, \n]
3        [\n, 12 months, \n]
4        [\n, 12 months, \n]
                ...         
29475                    NaN
29476    [\n, 13 months, \n]
29477    [\n, 12 months, \n]
29478                    NaN
29479                    NaN
Name: Warranty, Length: 29480, dtype: object

In [32]:
df['Warranty']=df['Warranty'].apply(lambda item: ",".join(item) if type(item) == list else item)
df['Warranty']

0        \n, 24 months, \n
1        \n, 24 months, \n
2        \n, 12 months, \n
3        \n, 12 months, \n
4        \n, 12 months, \n
               ...        
29475                  NaN
29476    \n, 13 months, \n
29477    \n, 12 months, \n
29478                  NaN
29479                  NaN
Name: Warranty, Length: 29480, dtype: object

In [33]:
df['Warranty']=df['Warranty'].apply(lambda item: re.sub(r'[\n,]', '', item) if type(item) == str else item)
df['Warranty']

0         24 months 
1         24 months 
2         12 months 
3         12 months 
4         12 months 
            ...     
29475            NaN
29476     13 months 
29477     12 months 
29478            NaN
29479            NaN
Name: Warranty, Length: 29480, dtype: object

In [34]:
df['Warranty'] = df['Warranty'].apply(lambda item: re.sub(r'[ \n, a-zA-Z ]','', item) if type(item) == str else item)
df['Warranty']

0         24
1         24
2         12
3         12
4         12
        ... 
29475    NaN
29476     13
29477     12
29478    NaN
29479    NaN
Name: Warranty, Length: 29480, dtype: object

In [35]:
#df1['Warranty']=df1['Warranty'].str.extract(r'(\d+\,?\d*)')
#df1['Warranty']=df1['Warranty'].str.replace(',',"").astype(float)
#df1['Warranty']

#For the time being "Warranty''s type is object if we want to make it float, here is the code!!!!!!!!!!!!

** <b>Column name of 'Mileage'<b>

In [36]:
df['Mileage']

0        120,200 km
1          5,000 km
2         18,900 km
3         18,995 km
4        197,000 km
            ...    
29475    145,000 km
29476     96,000 km
29477     94,343 km
29478    230,047 km
29479           NaN
Name: Mileage, Length: 29480, dtype: object

In [37]:
df['Mileage']=df['Mileage'].str.extract(r'(\d+\,?\d*)')
df['Mileage']

0        120,200
1          5,000
2         18,900
3         18,995
4        197,000
          ...   
29475    145,000
29476     96,000
29477     94,343
29478    230,047
29479        NaN
Name: Mileage, Length: 29480, dtype: object

In [38]:
df['Mileage']=df['Mileage'].str.replace(',',"").astype(float)
df['Mileage']

0        120200.0
1          5000.0
2         18900.0
3         18995.0
4        197000.0
           ...   
29475    145000.0
29476     96000.0
29477     94343.0
29478    230047.0
29479         NaN
Name: Mileage, Length: 29480, dtype: float64

** <b>Column name of 'First_Regeistration'<b>

In [39]:
df['First_Registration']

0        06/2016
1        06/2022
2        07/2020
3        01/2020
4        09/2015
          ...   
29475    06/2004
29476    04/2011
29477    11/2017
29478    07/2002
29479        NaN
Name: First_Registration, Length: 29480, dtype: object

In [40]:
df['First_Registration'] = pd.to_datetime(df['First_Registration'], format='%m/%Y')
df['First_Registration']

0       2016-06-01
1       2022-06-01
2       2020-07-01
3       2020-01-01
4       2015-09-01
           ...    
29475   2004-06-01
29476   2011-04-01
29477   2017-11-01
29478   2002-07-01
29479          NaT
Name: First_Registration, Length: 29480, dtype: datetime64[ns]

** <b>Column name of 'Gearbox'<b>

In [41]:
df['Gearbox']

0           [\nManual\n]
1        [\nAutomatic\n]
2        [\nAutomatic\n]
3        [\nAutomatic\n]
4        [\nAutomatic\n]
              ...       
29475    [\nAutomatic\n]
29476       [\nManual\n]
29477    [\nAutomatic\n]
29478    [\nAutomatic\n]
29479                NaN
Name: Gearbox, Length: 29480, dtype: object

In [42]:
df['Gearbox'] = df['Gearbox'].apply(lambda item: ",".join(item) if type(item) == list else item)
df['Gearbox']

0           \nManual\n
1        \nAutomatic\n
2        \nAutomatic\n
3        \nAutomatic\n
4        \nAutomatic\n
             ...      
29475    \nAutomatic\n
29476       \nManual\n
29477    \nAutomatic\n
29478    \nAutomatic\n
29479              NaN
Name: Gearbox, Length: 29480, dtype: object

In [43]:
df['Gearbox'] = df['Gearbox'].apply(lambda item: re.sub(r'[\n,]', '', item) if type(item) == str else item)
df['Gearbox']

0           Manual
1        Automatic
2        Automatic
3        Automatic
4        Automatic
           ...    
29475    Automatic
29476       Manual
29477    Automatic
29478    Automatic
29479          NaN
Name: Gearbox, Length: 29480, dtype: object

** <b>Column name of 'Fuel_Type'<b>

In [44]:
df1['Fuel_Type'].value_counts(dropna=False)

Gasoline                                                                                              8532
Diesel                                                                                                5911
Super 95                                                                                              3557
Diesel (Particle filter)                                                                              2816
NaN                                                                                                   2637
Regular/Benzine 91                                                                                    2065
Super E10 95                                                                                          1016
Regular/Benzine 91 (Particle filter)                                                                   555
Super 95 (Particle filter)                                                                             537
Super E10 95 (Particle filter)       

In [45]:
fuel_types = (['Diesel', 'Electric', 'Gasoline', 'Regular/Benzine 91',
       'Regular/Benzine 91 (Particle filter)', 'Super 95', 'Super E10 95',
       'LPG', 'Diesel (Particle filter)', 'Gasoline (Particle filter)',
       'Super 95 (Particle filter)', 'Liquid petroleum gas (LPG)',
       'Super Plus 98', 'Super E10 95 (Particle filter)',
       'Regular/Benzine E10 91', 'Super Plus E10 98',
       'Super Plus E10 98 (Particle filter)', 'Others',
       'Super Plus 98 (Particle filter)',
       'Liquid petroleum gas (LPG) / Super E10 95 / Regular/Benzine 91 / Super 95 / Super Plus 98 / Biogas',
       'Ethanol', 'Regular/Benzine E10 91 (Particle filter)', 'CNG',
       'Liquid petroleum gas (LPG) (Particle filter) / Super 95 / Super E10 95',
       'Others (Particle filter)',
       'Liquid petroleum gas (LPG) / Super 95 / Super E10 95',
       'LPG (Particle filter)', 'Electric (Particle filter)',
       'Domestic gas L',
       'Liquid petroleum gas (LPG) / Super 95 / Super Plus 98 / Super Plus E10 98 / Super E10 95',
       'CNG (Particle filter)', 'Domestic gas H',
       'Domestic gas L (Particle filter)', 'Biogas',
       'Domestic gas H / Super E10 95 / Super Plus E10 98 / Super 95 / Super Plus 98 / Domestic gas L',
       'Domestic gas L / Super 95 / Domestic gas H',
       'Super 95 / Super Plus 98 / Liquid petroleum gas (LPG)',
       'Liquid petroleum gas (LPG) (Particle filter)',
       'Liquid petroleum gas (LPG) / Super 95',
       'Liquid petroleum gas (LPG) / Super 95 / Super E10 95 / Super Plus 98',
       'Liquid petroleum gas (LPG) / Super 95 / Super Plus 98',
       'Liquid petroleum gas (LPG) / Super E10 95',
       'Liquid petroleum gas (LPG) / Super E10 95 / Super Plus E10 98 / Super Plus 98 / Super 95',
       'Hydrogen', 'Super 95 (Particle filter) / Super E10 95 / Ethanol',
       'Super 95 / Liquid petroleum gas (LPG)'])



In [46]:
def categorize_fuel_type(fuel_type):
    if isinstance(fuel_type, str):
        if re.search(r'LPG|Domestic|Hydrogen|Super', fuel_type):
            return 'LPG'
        elif re.search(r'Regular|Benzine|Super|Gasoline|Cng|Ethanol', fuel_type):
            return 'Gasoline'
        elif re.search(r'Diesel', fuel_type):
            return 'Diesel'
        elif re.search(r'Electric', fuel_type):
            return 'Electric'
        elif re.search(r'Others', fuel_type):
            return 'Others'
        else:
            return np.nan
    return fuel_type

In [47]:
df['Fuel_Type'] = df['Fuel_Type'].apply(lambda x: categorize_fuel_type(str(x)))
df['Fuel_Type']

0          Diesel
1        Electric
2        Gasoline
3        Gasoline
4          Diesel
           ...   
29475    Gasoline
29476      Diesel
29477    Gasoline
29478         LPG
29479         NaN
Name: Fuel_Type, Length: 29480, dtype: object

In [48]:
df['Fuel_Type'].unique()

array(['Diesel', 'Electric', 'Gasoline', nan, 'LPG', 'Others'],
      dtype=object)

** <b>Column name of 'Colour'<b>

In [49]:
df['Colour']

0         Black
1           NaN
2          Grey
3        Yellow
4         White
          ...  
29475      Grey
29476      Grey
29477      Blue
29478      Grey
29479       NaN
Name: Colour, Length: 29480, dtype: object

In [50]:
df['Colour'].unique()

array(['Black', nan, 'Grey', 'Yellow', 'White', 'Silver', 'Red', 'Blue',
       'Green', 'Beige', 'Brown', 'Gold', 'Bronze', 'Violet', 'Orange'],
      dtype=object)

** <b>Column name of 'Point'<b>

In [51]:
df['Paint']

0        Metallic
1             NaN
2             NaN
3             NaN
4        Metallic
           ...   
29475         NaN
29476    Metallic
29477    Metallic
29478         NaN
29479         NaN
Name: Paint, Length: 29480, dtype: object

In [52]:
df['Paint'].unique()

array(['Metallic', nan, 'Uni/basic'], dtype=object)

** <b>Column name of 'Desc'<b>

In [53]:
df['Desc'].head(5)

0    [ , Precio al contado: 16950 euros,  ,  , Vehí...
1    [ , Precio al contado: 88900 euros,  , AUTO OJ...
2    [Gris MANUFAKTUR mountaingrau magno,  , Linea/...
3    [ , Precio al contado: 48990 euros,  ,  , BONI...
4    [ , Precio al contado: 17400 euros,  , Vehícul...
Name: Desc, dtype: object

In [54]:
df['Desc'] = df['Desc'].apply(lambda x : "".join(x).strip() if type(x) == list else x).str.title()
df['Desc'].head(5)

0    Precio Al Contado: 16950 Euros  Vehículo Con H...
1    Precio Al Contado: 88900 Euros Auto Oja S.A. -...
2    Gris Manufaktur Mountaingrau Magno Linea/Estil...
3    Precio Al Contado: 48990 Euros  Bonificacion P...
4    Precio Al Contado: 17400 Euros Vehículo En Per...
Name: Desc, dtype: object

** <b>Column name of 'Seller'<b>

In [55]:
df['Seller']

0        Dealer
1        Dealer
2        Dealer
3        Dealer
4        Dealer
          ...  
29475    Dealer
29476    Dealer
29477    Dealer
29478    Dealer
29479       NaN
Name: Seller, Length: 29480, dtype: object

In [56]:
df['Seller'].unique()

array(['Dealer', 'Private seller', nan], dtype=object)

** <b>Column name of 'Seats'<b>

In [57]:
df['Seats']

0                NaN
1        [\n, 5, \n]
2        [\n, 5, \n]
3        [\n, 5, \n]
4        [\n, 5, \n]
            ...     
29475    [\n, 5, \n]
29476    [\n, 4, \n]
29477    [\n, 5, \n]
29478    [\n, 5, \n]
29479            NaN
Name: Seats, Length: 29480, dtype: object

In [58]:
df['Seats']=df['Seats'].apply(lambda item: ",".join(item) if type(item) == list else item)
df['Seats']

0              NaN
1        \n, 5, \n
2        \n, 5, \n
3        \n, 5, \n
4        \n, 5, \n
           ...    
29475    \n, 5, \n
29476    \n, 4, \n
29477    \n, 5, \n
29478    \n, 5, \n
29479          NaN
Name: Seats, Length: 29480, dtype: object

In [59]:
df['Seats'] = df['Seats'].apply(lambda item: re.sub(r'[\n,]', '', item) if type(item) == str else item)
df['Seats'] #float type yapmistim sonra tekrardan object yaptim ama en bastan calistirmam lazim ki duzelsin

0        NaN
1         5 
2         5 
3         5 
4         5 
        ... 
29475     5 
29476     4 
29477     5 
29478     5 
29479    NaN
Name: Seats, Length: 29480, dtype: object

** <b>Column name of 'Power'<b>

In [60]:
df['Power']

0                          NaN
1        [\n215 kW (292 hp)\n]
2        [\n310 kW (421 hp)\n]
3        [\n225 kW (306 hp)\n]
4        [\n100 kW (136 hp)\n]
                 ...          
29475    [\n125 kW (170 hp)\n]
29476     [\n84 kW (114 hp)\n]
29477    [\n187 kW (254 hp)\n]
29478    [\n147 kW (200 hp)\n]
29479                      NaN
Name: Power, Length: 29480, dtype: object

In [61]:
df['Power'] = df['Power'].map(lambda x : x[0].strip("\n ") if type(x) == list else x).str.extract(".*\((\d+.*\d*) ")
df['Power']
# 1 kW is nearly equal to 1.36 hp

0        NaN
1        292
2        421
3        306
4        136
        ... 
29475    170
29476    114
29477    254
29478    200
29479    NaN
Name: Power, Length: 29480, dtype: object

In [62]:
df['Power'].unique()

array([nan, '292', '421', '306', '136', '122', '190', '139', '90', '381',
       '211', '116', '109', '160', '163', '194', '360', '218', '170',
       '150', '235', '95', '557', '367', '639', '435', '523', '368',
       '362', '275', '510', '379', '224', '640', '102', '179', '177',
       '165', '156', '310', '91', '107', '137', '193', '152', '166',
       '118', '110', '265', '185', '140', '184', '320', '476', '340',
       '333', '204', '272', '571', '387', '530', '612', '522', '390',
       '315', '245', '326', '231', '252', '585', '517', '462', '500',
       '330', '469', '455', '238', '258', '188', '197', '158', '286',
       '296', '105', '525', '457', '394', '556', '276', '201', '199',
       '470', '86', '106', '230', '487', '481', '408', '299', '389',
       '730', '220', '200', '799', '630', '143', '489', '544', '250',
       '507', '449', '135', '228', '189', '658', '401', '239', '209',
       '237', '129', '514', '280', '503', '279', '442', '132', '75',
       '241', '69', 

** <b>Column name of 'Engine_Size'<b>

In [63]:
df['Engine_Size']

0                   NaN
1            [\n1 cc\n]
2        [\n1,991 cc\n]
3        [\n1,991 cc\n]
4        [\n2,143 cc\n]
              ...      
29475    [\n2,435 cc\n]
29476    [\n1,560 cc\n]
29477    [\n1,969 cc\n]
29478    [\n2,922 cc\n]
29479               NaN
Name: Engine_Size, Length: 29480, dtype: object

In [64]:
df['Engine_Size'] = df['Engine_Size'].map(lambda x : x[0].strip("\n c").replace(",",".") if type(x) == list else x).astype(float).round(1)
df['Engine_Size'].sample(5)

18280    4.0
17914    1.8
26704    2.0
24705    NaN
3037     1.4
Name: Engine_Size, dtype: float64

** <b>Column name of 'Gears'<b>

In [65]:
df['Gears']

0            NaN
1        [\n9\n]
2        [\n8\n]
3        [\n7\n]
4        [\n7\n]
          ...   
29475    [\n5\n]
29476    [\n6\n]
29477    [\n8\n]
29478        NaN
29479        NaN
Name: Gears, Length: 29480, dtype: object

In [66]:
df['Gears'] = df['Gears'].apply(lambda x : x[0].strip("\n") if type(x) == list else x)
df['Gears'].sample(5)

22774    NaN
18968      6
25908      6
11444      5
14238    NaN
Name: Gears, dtype: object

In [67]:
df['Gears'].unique()

array([nan, '9', '8', '7', '6', '1', '5', '4', '3', '0', '2', '10'],
      dtype=object)

** <b>Column name of 'Cylinders'<b>

In [68]:
df['Cylinders'] = df['Cylinders'].apply(lambda x : x[0].strip("\n") if type(x) == list else x)
df['Cylinders']

0        NaN
1        NaN
2          4
3          4
4          4
        ... 
29475      5
29476      4
29477      4
29478      6
29479    NaN
Name: Cylinders, Length: 29480, dtype: object

** <b>Column name of 'Fuel_Consumption'<b>

* This column's values were in lists, I extract first list that lists which are comb values. And than I took the fuel consumption in 100 km from the list as float type. 

In [69]:
df['Fuel_Consumption']

0                                                      NaN
1                                                      NaN
2        [[8.4 l/100 km (comb.)], [10.9 l/100 km (city)...
3        [[7.3 l/100 km (comb.)], [9.3 l/100 km (city)]...
4        [[4.9 l/100 km (comb.)], [5.7 l/100 km (city)]...
                               ...                        
29475    [[9.1 l/100 km (comb.)], [13.2 l/100 km (city)...
29476    [[3.8 l/100 km (comb.)], [4.3 l/100 km (city)]...
29477    [[6.5 l/100 km (comb.)], [8.6 l/100 km (city)]...
29478    [[10.4 l/100 km (comb.)], [15.3 l/100 km (city...
29479                                                  NaN
Name: Fuel_Consumption, Length: 29480, dtype: object

In [70]:
df['Fuel_Consumption'][2]

[['8.4 l/100 km (comb.)'],
 ['10.9 l/100 km (city)'],
 ['5.8 l/100 km (country)']]

In [71]:
df['Fuel_Consumption'] = df['Fuel_Consumption'].apply(lambda x: x[0] if isinstance(x, list) else None)
df['Fuel_Consumption']

0                           None
1                           None
2         [8.4 l/100 km (comb.)]
3         [7.3 l/100 km (comb.)]
4         [4.9 l/100 km (comb.)]
                  ...           
29475     [9.1 l/100 km (comb.)]
29476     [3.8 l/100 km (comb.)]
29477     [6.5 l/100 km (comb.)]
29478    [10.4 l/100 km (comb.)]
29479                       None
Name: Fuel_Consumption, Length: 29480, dtype: object

In [72]:
df['Fuel_Consumption'] = df['Fuel_Consumption'].apply(lambda x: x[0].split()[0] if isinstance(x, list) else None).astype(float)
df['Fuel_Consumption']

0         NaN
1         NaN
2         8.4
3         7.3
4         4.9
         ... 
29475     9.1
29476     3.8
29477     6.5
29478    10.4
29479     NaN
Name: Fuel_Consumption, Length: 29480, dtype: float64

** <b>Column name of 'Comfort_&_Convenience'<b>

In [73]:
df['Comfort_&_Convenience']

0                                                      NaN
1                                                      NaN
2                                                      NaN
3        [Air conditioning, Automatic climate control, ...
4        [Air conditioning, Electrical side mirrors, Mu...
                               ...                        
29475    [Air conditioning, Armrest, Automatic climate ...
29476    [Air conditioning, Armrest, Automatic climate ...
29477    [Air conditioning, Armrest, Automatic climate ...
29478    [Air conditioning, Armrest, Cruise control, El...
29479                                                  NaN
Name: Comfort_&_Convenience, Length: 29480, dtype: object

In [74]:
df["Comfort_&_Convenience"] = df["Comfort_&_Convenience"].apply(lambda x : x[0] if type(x) == list else x).str.title().astype("category")
df["Comfort_&_Convenience"].sample(5)

11818    Air Conditioning, Automatic Climate Control, C...
2954     Air Conditioning, Automatic Climate Control, C...
26752    Air Conditioning, Armrest, Automatic Climate C...
6335     Air Conditioning, Armrest, Automatic Climate C...
6606     Air Conditioning, Armrest, Cruise Control, Ele...
Name: Comfort_&_Convenience, dtype: category
Categories (14438, object): ['360° Camera', '360° Camera, Air Conditioning, Air Suspension..., '360° Camera, Air Conditioning, Air Suspension..., '360° Camera, Air Conditioning, Air Suspension..., ..., 'Sunroof, Tinted Windows', 'Sunroof, Wind Deflector', 'Tinted Windows', 'Wind Deflector']

In [75]:
lux_comf_equipment = ['Heads-Up Display', 'Panorama Roof', 'Massage Seats', 'Electrically Adjustable Seats']
mid_comf_equipment = ['Parking Assist System Sensors Front', 'Parking Assist System Sensors Rear', 'Cruise Control', 'Rain Sensor']
basic_comf_equipment = ['Park Distance Control', 'Fold Flat Passenger Seat', 'Wind Deflector', 'Sliding Door Right']


In [76]:
def categorize_car_comfort(equipment):
    if any(item in equipment for item in lux_comf_equipment):
        return 'Luxury'
    elif any(item in equipment for item in mid_comf_equipment):
        return 'Mid-Range'
    elif any(item in equipment for item in basic_comf_equipment):
        return 'Basic'
    else:
        return 'Uncategorized'

In [77]:
df['Com_Conv'] = df['Comfort_&_Convenience'].apply(lambda x: categorize_car_comfort(x))
df['Com_Conv'] 

0                  NaN
1                  NaN
2                  NaN
3            Mid-Range
4        Uncategorized
             ...      
29475        Mid-Range
29476        Mid-Range
29477        Mid-Range
29478        Mid-Range
29479              NaN
Name: Com_Conv, Length: 29480, dtype: object

In [78]:
df['Com_Conv'].value_counts(dropna=False)

Mid-Range        13991
Luxury            7260
NaN               4047
Uncategorized     3997
Basic              185
Name: Com_Conv, dtype: int64

** <b>Column name of 'Entertainment_&_Media'<b>

In [79]:
df['Entertainment_&_Media'].value_counts(dropna=False)

NaN                                                                                                                                                                          5836
[Bluetooth, USB]                                                                                                                                                              753
[Bluetooth]                                                                                                                                                                   665
[On-board computer]                                                                                                                                                           662
[CD player, On-board computer, Radio]                                                                                                                                         644
                                                                                                              

In [80]:
df["Entertainment_&_Media"] = df["Entertainment_&_Media"].apply(lambda x : x[0] if type(x) == list else x).str.title().astype("category")
df["Entertainment_&_Media"].sample(5)

25453    Android Auto, Apple Carplay, Bluetooth, Digita...
13498    Android Auto, Apple Carplay, Bluetooth, Cd Pla...
9427            Digital Cockpit, Digital Radio, Radio, Usb
15277                             On-Board Computer, Radio
7103                              On-Board Computer, Radio
Name: Entertainment_&_Media, dtype: category
Categories (1944, object): ['Android Auto', 'Android Auto, Apple Carplay', 'Android Auto, Apple Carplay, Bluetooth', 'Android Auto, Apple Carplay, Bluetooth, Cd Pl..., ..., 'Sound System, Usb', 'Sound System, Wlan / Wifi Hotspot', 'Usb', 'Wlan / Wifi Hotspot']

In [81]:
equipment_list = list(df["Entertainment_&_Media"].dropna().unique())

equipment_unique_list = {}

for equipment in equipment_list:
    items = equipment.split(', ')
    for item in items:
        item = item.strip()
        if item in equipment_unique_list:
            equipment_unique_list[item] += 1
        else:
            equipment_unique_list[item] = 1

equipment_unique_list

{'Usb': 1318,
 'Bluetooth': 1474,
 'Cd Player': 697,
 'Mp3': 844,
 'On-Board Computer': 1437,
 'Radio': 1404,
 'Digital Cockpit': 654,
 'Hands-Free Equipment': 1279,
 'Sound System': 927,
 'Digital Radio': 1017,
 'Android Auto': 838,
 'Apple Carplay': 904,
 'Induction Charging For Smartphones': 427,
 'Integrated Music Streaming': 483,
 'Wlan / Wifi Hotspot': 402,
 'Television': 92}

In [82]:
lux_media_equipment = ['Digital Cockpit', 'Integrated Music Streaming', 'Television']
mid_media_equipment = ['Usb', 'Bluetooth', 'On-Board Computer', 'Sound System']
basic_media_equipment = ['Cd Player', 'Mp3', 'Radio']

In [83]:
def categorize_car_media(equipment):
    if any(item in equipment for item in lux_media_equipment):
        return 'Luxury'
    elif any(item in equipment for item in mid_media_equipment):
        return 'Mid-Range'
    elif any(item in equipment for item in basic_media_equipment):
        return 'Basic'
    else:
        return 'Uncategorized'


In [84]:
df['Ent_Media'] = df['Entertainment_&_Media'].apply(lambda x: categorize_car_media(x))
df['Ent_Media'].sample(5)

10803       Luxury
10339    Mid-Range
8682        Luxury
24723    Mid-Range
25940    Mid-Range
Name: Ent_Media, dtype: object

In [85]:
df['Ent_Media'].value_counts(dropna=False)

Mid-Range        16411
Luxury            5853
NaN               5836
Basic             1308
Uncategorized       72
Name: Ent_Media, dtype: int64

** <b>Column name of 'Safety_&_Security'<b>

In [86]:
df["Safety_&_Security"].value_counts(dropna=False)

NaN                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            4065
[Isofix]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            

In [87]:
df["Safety_&_Security"] = df["Safety_&_Security"].apply(lambda x : x[0] if type(x) == list else x).str.title().astype("category")
df["Safety_&_Security"].sample(5)

19570    Abs, Alarm System, Daytime Running Lights, Dri...
4124     Abs, Adaptive Cruise Control, Adaptive Headlig...
13553    Abs, Central Door Lock, Daytime Running Lights...
8220     Abs, Adaptive Headlights, Driver-Side Airbag, ...
28051                 Driver-Side Airbag, Emergency System
Name: Safety_&_Security, dtype: category
Categories (12549, object): ['Abs', 'Abs, Adaptive Cruise Control, Adaptive Headli..., 'Abs, Adaptive Cruise Control, Adaptive Headli..., 'Abs, Adaptive Cruise Control, Adaptive Headli..., ..., 'Speed Limit Control System', 'Tire Pressure Monitoring System', 'Traction Control', 'Xenon Headlights']

In [88]:
sequrity_equipment_list = list(df["Safety_&_Security"].dropna().unique())

equipment_unique_list = {}

for equipment in sequrity_equipment_list:
    items = equipment.split(', ')
    for item in items:
        item = item.strip()
        if item in equipment_unique_list:
            equipment_unique_list[item] += 1
        else:
            equipment_unique_list[item] = 1

equipment_unique_list

{'Abs': 11613,
 'Central Door Lock': 9744,
 'Driver-Side Airbag': 11337,
 'Electronic Stability Control': 10588,
 'Traction Control': 8249,
 'Isofix': 8983,
 'Xenon Headlights': 2249,
 'Bi-Xenon Headlights': 1545,
 'Fog Lights': 8017,
 'Head Airbag': 5229,
 'Immobilizer': 9041,
 'Passenger-Side Airbag': 11209,
 'Power Steering': 10700,
 'Side Airbag': 10475,
 'Speed Limit Control System': 4341,
 'Emergency Brake Assistant': 5903,
 'Alarm System': 3729,
 'Central Door Lock With Remote Control': 5308,
 'Tire Pressure Monitoring System': 8580,
 'Daytime Running Lights': 7135,
 'Led Daytime Running Lights': 6867,
 'Led Headlights': 5056,
 'Lane Departure Warning System': 4958,
 'Traffic Sign Recognition': 4010,
 'Full-Led Headlights': 1819,
 'Distance Warning System': 3604,
 'Driver Drowsiness Detection': 3470,
 'Emergency System': 3462,
 'High Beam Assist': 2905,
 'Rear Airbag': 2041,
 'Adaptive Cruise Control': 2468,
 'Blind Spot Monitor': 2353,
 'Glare-Free High Beam Headlights': 1053,


In [89]:
lux_sequrity_features = ['Full-Led Headlights', 'Adaptive Cruise Control', 'Lane Departure Warning System', 'Night View Assist']
mid_sequrity_features = ['Abs', 'Electronic Stability Control', 'Isofix', 'Tire Pressure Monitoring System']
basic_sequrity_features = ['Fog Lights', 'Immobilizer', 'Speed Limit Control System', 'Emergency Brake Assistant']

In [90]:
def categorize_car_sequrity(equipment):
    if any(item in equipment for item in lux_sequrity_features):
        return 'Luxury'
    elif any(item in equipment for item in mid_sequrity_features):
        return 'Mid-Range'
    elif any(item in equipment for item in basic_sequrity_features):
        return 'Basic'
    else:
        return 'Uncategorized'

In [91]:
df['Safe_Secure'] = df['Safety_&_Security'].apply(lambda x: categorize_car_sequrity(x))
df['Safe_Secure'].sample(5)

17659          NaN
21713       Luxury
5469     Mid-Range
6320     Mid-Range
9577     Mid-Range
Name: Safe_Secure, dtype: object

In [92]:
df['Safe_Secure'].value_counts(dropna=False)

Mid-Range        15145
Luxury            9454
NaN               4065
Uncategorized      453
Basic              363
Name: Safe_Secure, dtype: int64

** <b>Column name of 'Extras'<b>

In [93]:
df['Extras'].value_counts(dropna=False)

NaN                                                                                                                                                                                                                                                                   6000
[Alloy wheels]                                                                                                                                                                                                                                                        3408
[Alloy wheels, Roof rack]                                                                                                                                                                                                                                              380
[Alloy wheels, Sport seats]                                                                                                                                                                            

In [94]:
df["Extras"] = df["Extras"].apply(lambda x : x[0] if type(x) == list else x).str.title().astype("category")
df["Extras"].sample(5)

12531    Alloy Wheels (18"), Ambient Lighting, Automati...
16291                              Roof Rack, Steel Wheels
17539           Cargo Barrier, Touch Screen, Voice Control
20246    Alloy Wheels (17"), E10-Enabled, Electronic Pa...
3558                       Alloy Wheels (17"), Sport Seats
Name: Extras, dtype: category
Categories (9360, object): ['All Season Tyres', 'All Season Tyres, Alloy Wheels', 'All Season Tyres, Alloy Wheels (13"), Catalyt..., 'All Season Tyres, Alloy Wheels (14")', ..., 'Voice Control, Winter Tyres', 'Winter Package', 'Winter Package, Winter Tyres', 'Winter Tyres']

In [95]:
equipment_list = list(df["Extras"].dropna().unique())

equipment_unique_list = {}

for equipment in equipment_list:
    items = equipment.split(', ')
    for item in items:
        item = item.strip()
        if item in equipment_unique_list:
            equipment_unique_list[item] += 1
        else:
            equipment_unique_list[item] = 1

equipment_unique_list

{'Sport Seats': 4013,
 'Alloy Wheels (17")': 1027,
 'Catalytic Converter': 1576,
 'Emergency Tyre Repair Kit': 3328,
 'Headlight Washer System': 1993,
 "Smoker'S Package": 518,
 'Sport Package': 2521,
 'Alloy Wheels': 4365,
 'Spoiler': 1388,
 'Sport Suspension': 2510,
 'Automatically Dimming Interior Mirror': 5140,
 'Voice Control': 4770,
 'Cargo Barrier': 2061,
 'Emergency Tyre': 1148,
 'Summer Tyres': 3746,
 'Winter Package': 2492,
 'Alloy Wheels (19")': 681,
 'Ski Bag': 564,
 'Winter Tyres': 957,
 'Electronic Parking Brake': 2032,
 'Shift Paddles': 2455,
 'Touch Screen': 5068,
 'Sliding Door': 60,
 'Alloy Wheels (18")': 933,
 'Ambient Lighting': 2840,
 'Alloy Wheels (16")': 683,
 'E10-Enabled': 2271,
 'Spare Tyre': 1226,
 'Alloy Wheels (21")': 59,
 'Trailer Hitch': 1621,
 'Alloy Wheels (20")': 243,
 'Steel Wheels': 686,
 'Biodiesel Conversion': 8,
 'Handicapped Enabled': 26,
 'Tuned Car': 235,
 'Range Extender': 42,
 'All Season Tyres': 991,
 'Alloy Wheels (14")': 37,
 'Alloy Wheels

In [96]:
lux_extra_features = ['Sport Seats', 'Headlight Washer System', 'Automatically Dimming Interior Mirror', 'Ambient Lighting']
mid_extra_features = ['Sport Suspension', 'Alloy Wheels (18")', 'Electronic Parking Brake', 'Roof Rack']
basic_extra_features = ['Catalytic Converter', 'Emergency Tyre Repair Kit', 'Smoker\'s Package', 'Winter Package']

In [97]:
def categorize_car_extra(equipment):
    if any(item in equipment for item in lux_extra_features):
        return 'Luxury'
    elif any(item in equipment for item in mid_extra_features):
        return 'Mid-Range'
    elif any(item in equipment for item in basic_extra_features):
        return 'Basic'
    else:
        return 'Uncategorized'

In [98]:
df['Extras2'] = df['Extras'].apply(lambda x: categorize_car_extra(x))
df['Extras2']

0                  NaN
1                  NaN
2                  NaN
3               Luxury
4                  NaN
             ...      
29475    Uncategorized
29476           Luxury
29477           Luxury
29478           Luxury
29479              NaN
Name: Extras2, Length: 29480, dtype: object

In [99]:
df['Extras2'].value_counts(dropna=False)

Luxury           12129
Uncategorized     6646
NaN               6000
Mid-Range         3082
Basic             1623
Name: Extras2, dtype: int64

** <b>Column name of 'Empty_Weight'<b>

In [100]:
df['Empty_Weight']

0                   NaN
1                   NaN
2                   NaN
3                   NaN
4        [\n1,545 kg\n]
              ...      
29475    [\n1,357 kg\n]
29476    [\n1,356 kg\n]
29477    [\n1,617 kg\n]
29478    [\n1,573 kg\n]
29479               NaN
Name: Empty_Weight, Length: 29480, dtype: object

In [101]:
df['Empty_Weight'] = df['Empty_Weight'].map(lambda x: x[0].strip('\n kg').replace(',','') if type(x) == list else x).astype(float)
df['Empty_Weight'].sample(5)

20758    1460.0
20466       NaN
1177        NaN
6156        NaN
17428     950.0
Name: Empty_Weight, dtype: float64

** <b>Column name of 'Model_Code'<b>

In [102]:
df['Model_Code'].value_counts()

[\n, 8212/AFJ, \n]    75
[\n, 1727/AAM, \n]    64
[\n, 1349/AGI, \n]    61
[\n, 1889/ABU, \n]    55
[\n, 7104/AAL, \n]    53
                      ..
[\n, 3333/BJK, \n]     1
[\n, 3333/BHJ, \n]     1
[\n, 3333/BPA, \n]     1
[\n, 3333/BFV, \n]     1
[\n, 9101/449, \n]     1
Name: Model_Code, Length: 2186, dtype: int64

In [103]:
df['Model_Code'] = df['Model_Code'].map(lambda x: x[0].strip('\n ').replace(',','') if type(x) == list else x)
df['Model_Code'].sample(5)

12212          NaN
22010          NaN
12424          NaN
3746      1844/AKH
15504          NaN
Name: Model_Code, dtype: object

** <b>Column name of 'General_Inspection'<b>

In [104]:
df['General_Inspection']

0            NaN
1            NaN
2            NaN
3            NaN
4            NaN
          ...   
29475    08/2023
29476        NaN
29477        New
29478    05/2023
29479        NaN
Name: General_Inspection, Length: 29480, dtype: object

*The errors='coerce' argument handles any invalid or missing dates by converting them to NaT values.

In [105]:
df['General_Inspection'] = pd.to_datetime(df['General_Inspection'], errors='coerce')
df['General_Inspection']

0              NaT
1              NaT
2              NaT
3              NaT
4              NaT
           ...    
29475   2023-08-01
29476          NaT
29477          NaT
29478   2023-05-01
29479          NaT
Name: General_Inspection, Length: 29480, dtype: datetime64[ns]

** <b>Column name of 'Previous_Owner'<b>

In [106]:
df['Previous_Owner']

0                               NaN
1                               NaN
2                               NaN
3                               NaN
4                               NaN
                    ...            
29475                           NaN
29476     [[96,000 km, 04/2011], 1]
29477                           NaN
29478    [[230,047 km, 07/2002], 5]
29479                           NaN
Name: Previous_Owner, Length: 29480, dtype: object

In [107]:
df['Mileage']

0        120200.0
1          5000.0
2         18900.0
3         18995.0
4        197000.0
           ...   
29475    145000.0
29476     96000.0
29477     94343.0
29478    230047.0
29479         NaN
Name: Mileage, Length: 29480, dtype: float64

In [108]:
df['Previous_Owner'] = df['Previous_Owner'].apply(lambda x: x[1] if isinstance(x, list) else None).astype(float)
df['Previous_Owner']

0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
        ... 
29475    NaN
29476    1.0
29477    NaN
29478    5.0
29479    NaN
Name: Previous_Owner, Length: 29480, dtype: float64

** <b>Column name of 'Upholstery'<b>

In [109]:
df['Upholstery'].value_counts(dropna=False)

NaN             10870
Cloth            8736
Full leather     5439
Part leather     2835
alcantara         764
Other             628
Velour            208
Name: Upholstery, dtype: int64

In [110]:
df['Upholstery'] = df['Upholstery'].str.lower().str.replace(' ', '_')
df['Upholstery'].value_counts(dropna=False)

NaN             10870
cloth            8736
full_leather     5439
part_leather     2835
alcantara         764
other             628
velour            208
Name: Upholstery, dtype: int64

** <b>Column name of 'Emission_Class'<b>

In [111]:
df['Emission_Class'].value_counts(dropna=False)

NaN             11621
Euro 6           6418
Euro 6d-TEMP     3399
Euro 6d          2858
Euro 5           2389
Euro 4           1743
Euro 3            523
Euro 2            217
Euro 1            172
Euro 6c           140
Name: Emission_Class, dtype: int64

In [112]:
emission = {
    'Euro 6':'e_60',       
    'Euro 6d-TEMP':'e_66',
    'Euro 6d':'e_64', 
    'Euro 5 ':'e_50',
    'Euro 4':'e_40',
    'Euro 3':'e_30',
    'Euro 2':'e_20',
    'Euro 1':'e_10,',
    'Euro 6c':'e_62'
}

In [113]:
df['Emission_Class'] = df['Emission_Class'].map(emission)
df['Emission_Class']

0         NaN
1         NaN
2         NaN
3         NaN
4         NaN
         ... 
29475    e_40
29476     NaN
29477    e_60
29478    e_30
29479     NaN
Name: Emission_Class, Length: 29480, dtype: object

In [114]:
df['Emission_Class'].unique()

array([nan, 'e_40', 'e_60', 'e_66', 'e_64', 'e_62', 'e_30', 'e_20',
       'e_10,'], dtype=object)

In [115]:
df

Make_Model  \
0         Mercedes-Benz A 160   
1       Mercedes-Benz EQE 350   
2      Mercedes-Benz A 45 AMG   
3      Mercedes-Benz A 35 AMG   
4      Mercedes-Benz A 45 AMG   
...                       ...   
29475               Volvo S40   
29476               Volvo C30   
29477               Volvo S90   
29478               Volvo S80   
29479                     NaN   

                                       Short_Description           Make  \
0                                                    CDi  Mercedes-Benz   
1                                                   350+  Mercedes-Benz   
2                                       S 4Matic+ 8G-DCT  Mercedes-Benz   
3                                         4Matic+ 7G-DCT  Mercedes-Benz   
4                               200CDI BE Line 4M 7G-DCT  Mercedes-Benz   
...                                                  ...            ...   
29475        2.4i Momentum/ Airco/ NAP/ APK/ Elek. Ramen          Volvo   
29476                    1.6 D DRIVe Start/Stop R-design          Volvo   
29477  T5 AUT(8) MOMENTUM BUSINESS SCHUIFDAK 360GR CA...          Volvo   
29478  2.9 Executive G. NETTE AUTO! LEER! NAVI! CRUIS...          Volvo   
29479                                                NaN            NaN   

        Model Location    Price  Body_Type              Type Doors  \
0        A160       ES  16950.0   Compact              Used     5    
1      EQE350       ES  80900.0   Compact    Pre-registered     4    
2      A45AMG       ES  69900.0   Compact              Used     5    
3      A35AMG       ES  46990.0   Compact              Used     5    
4      A45AMG       ES  16800.0   Compact              Used     5    
...       ...      ...      ...        ...               ...   ...   
29475     S40       NL   5499.0     Sedan              Used     4    
29476     C30       BE   7300.0     Sedan              Used     3    
29477     S90       NL  29900.0     Sedan              Used     4    
29478     S80       NL   4440.0     Sedan              Used     4    
29479     NaN      NaN      NaN        NaN               NaN   NaN   

      Country_Version  ... Taxi_Or_Rental_Car Availability  \
0              Spain   ...                NaN          NaN   
1              Spain   ...                NaN          NaN   
2              Spain   ...                NaN          NaN   
3              Spain   ...                NaN          NaN   
4              Spain   ...                NaN          NaN   
...               ...  ...                ...          ...   
29475             NaN  ...                NaN          NaN   
29476        Belgium   ...                NaN          NaN   
29477             NaN  ...                NaN          NaN   
29478             NaN  ...                NaN          NaN   
29479             NaN  ...                NaN          NaN   

       Last_Timing_Belt_Change Electric_Range_(Wltp) Power_Consumption_(Wltp)  \
0                          NaN                   NaN                      NaN   
1                          NaN                   NaN                      NaN   
2                          NaN                   NaN                      NaN   
3                          NaN                   NaN                      NaN   
4                          NaN                   NaN                      NaN   
...                        ...                   ...                      ...   
29475                      NaN                   NaN                      NaN   
29476                      NaN                   NaN                      NaN   
29477                      NaN                   NaN                      NaN   
29478                      NaN                   NaN                      NaN   
29479                      NaN                   NaN                      NaN   

      Battery_Ownership       Com_Conv  Ent_Media Safe_Secure        Extras2  
0                   NaN            NaN        NaN         NaN            NaN  
1   

In [116]:
df.columns

Index(['Make_Model', 'Short_Description', 'Make', 'Model', 'Location', 'Price',
       'Body_Type', 'Type', 'Doors', 'Country_Version', 'Offer_Number',
       'Warranty', 'Mileage', 'First_Registration', 'Gearbox', 'Fuel_Type',
       'Colour', 'Paint', 'Desc', 'Seller', 'Seats', 'Power', 'Engine_Size',
       'Gears', 'Co₂-Emissions', 'Manufacturer_Colour', 'Drivetrain',
       'Cylinders', 'Fuel_Consumption', 'Comfort_&_Convenience',
       'Entertainment_&_Media', 'Safety_&_Security', 'Extras', 'Empty_Weight',
       'Model_Code', 'General_Inspection', 'Last_Service',
       'Full_Service_History', 'Non-Smoker_Vehicle', 'Emission_Class',
       'Emissions_Sticker', 'Upholstery_Colour', 'Upholstery',
       'Production_Date', 'Previous_Owner', 'Other_Fuel_Types',
       'Power_Consumption', 'Energy_Efficiency_Class', 'Co₂-Efficiency',
       'Fuel_Consumption_(Wltp)', 'Co₂-Emissions_(Wltp)', 'Available_From',
       'Taxi_Or_Rental_Car', 'Availability', 'Last_Timing_Belt_Change',
   

*In above, I forgotto delete 4 column which are name 'Comfort_&_Convenience',
       'Entertainment_&_Media', 'Safety_&_Security', 'Extras'\n
       
*We're going to delete them

In [117]:
df.drop(columns=['Comfort_&_Convenience', 'Entertainment_&_Media','Safety_&_Security', 'Extras'], inplace=True)

In [118]:
df.shape

(29480, 58)

# PART- 2 `( Handling With Missing Vales )`

## <p style="background-color:yellow; font-family:newtimeroman; color:blue; font-size:150%; text-align:center; border-radius:10px 10px;">Looking for percentage of null values</p> 

* If any column's percentage is over 70 we can drop that column
* If any column's percentage is betseen 50-70 we should try to fill that column, if we can't fill or 
if that colum is useless drop it
* If any column's percentage is below 50 we can't drop that column


In [119]:
df.isna().sum()

Make_Model                    850
Short_Description             850
Make                          850
Model                         850
Location                      850
Price                         850
Body_Type                     850
Type                          850
Doors                        1209
Country_Version             12591
Offer_Number                 6380
Warranty                    13696
Mileage                       851
First_Registration            852
Gearbox                      1098
Fuel_Type                    2733
Colour                       2574
Paint                       14985
Desc                         1433
Seller                        850
Seats                        3975
Power                        1422
Engine_Size                  2253
Gears                       10526
Co₂-Emissions               10886
Manufacturer_Colour          7693
Drivetrain                  12587
Cylinders                   10628
Fuel_Consumption             6095
Empty_Weight  

In [120]:
null_percentage = (df.isna().sum()/len(df))*100
null_percentage

Make_Model                   2.883311
Short_Description            2.883311
Make                         2.883311
Model                        2.883311
Location                     2.883311
Price                        2.883311
Body_Type                    2.883311
Type                         2.883311
Doors                        4.101085
Country_Version             42.710312
Offer_Number                21.641791
Warranty                    46.458616
Mileage                      2.886703
First_Registration           2.890095
Gearbox                      3.724559
Fuel_Type                    9.270692
Colour                       8.731343
Paint                       50.831072
Desc                         4.860923
Seller                       2.883311
Seats                       13.483718
Power                        4.823609
Engine_Size                  7.642469
Gears                       35.705563
Co₂-Emissions               36.926730
Manufacturer_Colour         26.095658
Drivetrain  

In [121]:
columns_to_drop = null_percentage[null_percentage > 60].index  
df.drop(columns_to_drop, axis=1, inplace=True)
df

Make_Model  \
0         Mercedes-Benz A 160   
1       Mercedes-Benz EQE 350   
2      Mercedes-Benz A 45 AMG   
3      Mercedes-Benz A 35 AMG   
4      Mercedes-Benz A 45 AMG   
...                       ...   
29475               Volvo S40   
29476               Volvo C30   
29477               Volvo S90   
29478               Volvo S80   
29479                     NaN   

                                       Short_Description           Make  \
0                                                    CDi  Mercedes-Benz   
1                                                   350+  Mercedes-Benz   
2                                       S 4Matic+ 8G-DCT  Mercedes-Benz   
3                                         4Matic+ 7G-DCT  Mercedes-Benz   
4                               200CDI BE Line 4M 7G-DCT  Mercedes-Benz   
...                                                  ...            ...   
29475        2.4i Momentum/ Airco/ NAP/ APK/ Elek. Ramen          Volvo   
29476                    1.6 D DRIVe Start/Stop R-design          Volvo   
29477  T5 AUT(8) MOMENTUM BUSINESS SCHUIFDAK 360GR CA...          Volvo   
29478  2.9 Executive G. NETTE AUTO! LEER! NAVI! CRUIS...          Volvo   
29479                                                NaN            NaN   

        Model Location    Price  Body_Type              Type Doors  \
0        A160       ES  16950.0   Compact              Used     5    
1      EQE350       ES  80900.0   Compact    Pre-registered     4    
2      A45AMG       ES  69900.0   Compact              Used     5    
3      A35AMG       ES  46990.0   Compact              Used     5    
4      A45AMG       ES  16800.0   Compact              Used     5    
...       ...      ...      ...        ...               ...   ...   
29475     S40       NL   5499.0     Sedan              Used     4    
29476     C30       BE   7300.0     Sedan              Used     3    
29477     S90       NL  29900.0     Sedan              Used     4    
29478     S80       NL   4440.0     Sedan              Used     4    
29479     NaN      NaN      NaN        NaN               NaN   NaN   

      Country_Version  ... Empty_Weight Full_Service_History  Emission_Class  \
0              Spain   ...          NaN                  NaN             NaN   
1              Spain   ...          NaN                  NaN             NaN   
2              Spain   ...          NaN                  NaN             NaN   
3              Spain   ...          NaN                  NaN             NaN   
4              Spain   ...       1545.0                  NaN             NaN   
...               ...  ...          ...                  ...             ...   
29475             NaN  ...       1357.0                  Yes            e_40   
29476        Belgium   ...       1356.0                  Yes             NaN   
29477             NaN  ...       1617.0                  Yes            e_60   
29478             NaN  ...       1573.0                  NaN            e_30   
29479             NaN  ...          NaN                  NaN             NaN   

      Upholstery_Colour    Upholstery Previous_Owner       Com_Conv  \
0                   NaN           NaN            NaN            NaN   
1                   NaN           NaN            NaN            NaN   
2                   NaN           NaN            NaN            NaN   
3                   NaN           NaN            NaN      Mid-Range   
4                   NaN           NaN            NaN  Uncategorized   
...                 ...           ...            ...            ...   
29475               NaN           NaN            NaN      Mid-Range   
29476             Black           NaN            1.0      Mid-Range   
29477             Other  full_leather            NaN      Mid-Range   
29478             Beige  full_leather            5.0      Mid-Range   
29479               NaN           NaN            NaN            NaN   

       Ent_Media Safe_Secure        Extras2  
0            NaN    

## <p style="background-color:yellow; font-family:newtimeroman; color:blue; font-size:150%; text-align:center; border-radius:10px 10px;">Checking if there are any row with full of null values</p> 

* If all values are null, I'll drop that row

In [122]:
df.iloc[29479]

Make_Model              NaN
Short_Description       NaN
Make                    NaN
Model                   NaN
Location                NaN
Price                   NaN
Body_Type               NaN
Type                    NaN
Doors                   NaN
Country_Version         NaN
Offer_Number            NaN
Warranty                NaN
Mileage                 NaN
First_Registration      NaT
Gearbox                 NaN
Fuel_Type               NaN
Colour                  NaN
Paint                   NaN
Desc                    NaN
Seller                  NaN
Seats                   NaN
Power                   NaN
Engine_Size             NaN
Gears                   NaN
Co₂-Emissions           NaN
Manufacturer_Colour     NaN
Drivetrain              NaN
Cylinders               NaN
Fuel_Consumption        NaN
Empty_Weight            NaN
Full_Service_History    NaN
Emission_Class          NaN
Upholstery_Colour       NaN
Upholstery              NaN
Previous_Owner          NaN
Com_Conv            

* As it appears, we have some rows which are full of null values
* We have 850 rows which are full of null values, when I drop them, df will have 28630 rows

In [123]:
df.dropna(axis=0, how='all', inplace=True)
df.reset_index(drop=True, inplace=True)

In [147]:
df['Power']=df['Power'].astype(object)

In [150]:
df.to_csv("cleaned.csv")

In [148]:
df['Power']

0        nan
1        292
2        421
3        306
4        136
        ... 
28625    392
28626    170
28627    114
28628    254
28629    200
Name: Power, Length: 28630, dtype: object

# PART- 3 `( Handling With Outliers )`

# Final Step (Checking final situation of data via graphs)

## Export dataframe to csv file (without dummy)

In [ ]:
df.to_csv("final_scout_not_dummy.csv", index=False)

# Dummy Operation

## Export dataframe to csv file (dummy)

In [ ]:
df_dummied.to_csv("final_scout_dummy.csv", index=False)